File: tools/packages/codeflow-utils-python/src/codeflow_utils/formatting/date.py
"""Date and time formatting utilities."""

from datetime import datetime
from typing import Optional


def format_datetime(dt: datetime, format_str: str = "%Y-%m-%d %H:%M:%S") -> str:
    """
    Format datetime to string.

    Args:
        dt: Datetime object
        format_str: Format string (default: ISO-like format)

    Returns:
        Formatted datetime string
    """
    return dt.strftime(format_str)


def format_iso_datetime(dt: datetime) -> str:
    """
    Format datetime to ISO 8601 string.

    Args:
        dt: Datetime object

    Returns:
        ISO 8601 formatted string
    """
    return dt.isoformat()


def format_relative_time(dt: datetime, now: Optional[datetime] = None) -> str:
    """
    Format datetime as relative time (e.g., "2 hours ago").

    Args:
        dt: Datetime object
        now: Current datetime (defaults to now)

    Returns:
        Relative time string
    """
    if now is None:
        now = datetime.now()

    delta = now - dt

    if delta.total_seconds() < 0:
        # Future time
        delta = dt - now
        if delta.days > 365:
            years = delta.days // 365
            return f"in {years} year{'s' if years > 1 else ''}"
        elif delta.days > 30:
            months = delta.days // 30
            return f"in {months} month{'s' if months > 1 else ''}"
        elif delta.days > 0:
            return f"in {delta.days} day{'s' if delta.days > 1 else ''}"
        elif delta.seconds > 3600:
            hours = delta.seconds // 3600
            return f"in {hours} hour{'s' if hours > 1 else ''}"
        elif delta.seconds > 60:
            minutes = delta.seconds // 60
            return f"in {minutes} minute{'s' if minutes > 1 else ''}"
        else:
            return "in a moment"

    # Past time
    if delta.days > 365:
        years = delta.days // 365
        return f"{years} year{'s' if years > 1 else ''} ago"
    elif delta.days > 30:
        months = delta.days // 30
        return f"{months} month{'s' if months > 1 else ''} ago"
    elif delta.days > 0:
        return f"{delta.days} day{'s' if delta.days > 1 else ''} ago"
    elif delta.seconds > 3600:
        hours = delta.seconds // 3600
        return f"{hours} hour{'s' if hours > 1 else ''} ago"
    elif delta.seconds > 60:
        minutes = delta.seconds // 60
        return f"{minutes} minute{'s' if minutes > 1 else ''} ago"
    else:
        return "just now"

File: desktop/packages/utils/src/formatting/date.ts
/**
 * Date and time formatting utilities.
 */

/**
 * Format datetime to string.
 *
 * @param date - Date object to format
 * @param format - Format string ('iso' for ISO 8601, or custom format)
 * @returns Formatted datetime string
 */
export function formatDateTime(date: Date, format: string = "default"): string {
  if (format === "iso") {
    return date.toISOString();
  }

  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  const hours = String(date.getHours()).padStart(2, "0");
  const minutes = String(date.getMinutes()).padStart(2, "0");
  const seconds = String(date.getSeconds()).padStart(2, "0");

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}

/**
 * Format datetime as relative time (e.g., "2 hours ago").
 *
 * @param date - Date object
 * @param now - Current date (defaults to now)
 * @returns Relative time string
 */
export function formatRelativeTime(date: Date, now: Date = new Date()): string {
  const delta = now.getTime() - date.getTime();
  const seconds = Math.floor(Math.abs(delta) / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  const months = Math.floor(days / 30);
  const years = Math.floor(days / 365);

  const isFuture = delta < 0;
  const prefix = isFuture ? "in " : "";
  const suffix = isFuture ? "" : " ago";

  if (years > 0) {
    return `${prefix}${years} year${years > 1 ? "s" : ""}${suffix}`;
  } else if (months > 0) {
    return `${prefix}${months} month${months > 1 ? "s" : ""}${suffix}`;
  } else if (days > 0) {
    return `${prefix}${days} day${days > 1 ? "s" : ""}${suffix}`;
  } else if (hours > 0) {
    return `${prefix}${hours} hour${hours > 1 ? "s" : ""}${suffix}`;
  } else if (minutes > 0) {
    return `${prefix}${minutes} minute${minutes > 1 ? "s" : ""}${suffix}`;
  } else {
    return isFuture ? "in a moment" : "just now";
  }
}
